hooks/bootstrap.py

```diff
@@ -0,0 +1,84 @@
+    the ToolkitManager's bootstrapping process.
+
+    The hook will be instantiated only after a configuration has been selected
+    by the ``ToolkitManager``. Therefore, the ``download_bundle`` method
+    will not be invoked to download a configuration. However, the Toolkit Core,
+    applications, frameworks and engines will be downloaded through the hook.
+    """
+
+    def init(self, shotgun, pipeline_configuration_id, configuration_descriptor):
+        """
+        This method is invoked when the hook is instantiated.
+
+        It can be used to cache information so it can be reused.
+
+        :param shotgun: Connection to the Shotgun site.
+        :type shotgun: :class:`shotgun_api3.Shotgun`
+
+        :param pipeline_configuration_id: Id of the pipeline configuration we're bootstrapping into.
+            If None, the ToolkitManager is bootstrapping into the base configuration.
+        :type pipeline_configuration_id: int
+
+        :param configuration_descriptor: Configuration the manager is bootstrapping into.
+        :type configuration_descriptor: :class:`sgtk.descriptor.ConfigDescriptor`
+        """
+        self.shotgun = shotgun
+        self.pipeline_configuration_id = pipeline_configuration_id
+        self.configuration_descriptor = configuration_descriptor
+
+    def download_bundle(self, descriptor):
+        """
+        This method is invoked during bootstrapping when downloading the core or
+        the bundles used by a configuration.
+
+        Your override must download the descriptor's content before the method
+        ends or the bundle cache will be left in an inconsistent state.
+
+        The safest way to download a bundle into the bundle cache is by using the
+        :meth:`_open_write_location` method with the `with` statement.
+
+        .. example::
+            with self._open_write_location(descriptor) as write_location:
+                # Put code that writes the bundle at write_location here.
+
+        Any exception raised in this method will abort the bootstrap process.
+
+        The default implementation will download the descriptor's content from
+        its source.
+
+        :param descriptor: The descriptor that will be downloaded.
+        :type descriptor: :class:`~sgtk.descriptor.Descriptor`
+        """
+        descriptor.download_local()
+
+    @contextlib.contextmanager
+    def _open_write_location(self, descriptor):
+        """
+        Allows to write a bundle to the primary bundle cache location.
+
+        If an exception is raised during the process, the files will be deleted
+        and the bundle cache will be left intact. Be careful to properly copy
+        all the files during the duration of the ``_open_write_location`` method
+        or the bundle cache will be left in an inconsistent state.
+        """
+        with descriptor._io_descriptor.open_write_location() as write_location:
```

Comment:
This doesn't seem right - we are accessing `descriptor._io_descriptor.open_write_location` which is an internal thing. perhaps we can just expose `open_write_location` as an official thing on `descriptor`?

Reply:
Well, I didn't expose it on the descriptor class because you had suggested that we should probably expose the functionality on the hook instead of making this part of the official descriptor API, so that's what I came up with.

Reply:
Or are you suggesting that we put the method on the descriptor object but exclude it from the doc?